internal/pkg/cli/file/watch.go

```diff
@@ -0,0 +1,130 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package file
+
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// RecursiveWatcher wraps an fsnotify Watcher to recursively watch all files in a directory.
+type RecursiveWatcher struct {
+	fsnotifyWatcher *fsnotify.Watcher
+	done            chan struct{}
+	closed          bool
+	events          chan fsnotify.Event
+	errors          chan error
+}
+
+// NewRecursiveWatcher returns a RecursiveWatcher which notifies when changes are made to files inside a recursive directory tree.
+func NewRecursiveWatcher() (*RecursiveWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	rw := &RecursiveWatcher{
+		events:          make(chan fsnotify.Event),
+		errors:          make(chan error),
+		fsnotifyWatcher: watcher,
+		done:            make(chan struct{}),
+		closed:          false,
+	}
+
+	go rw.start()
+
+	return rw, nil
+}
+
+// Add recursively adds a directory tree to the list of watched files.
+func (rw *RecursiveWatcher) Add(path string) error {
+	if rw.closed {
+		return fsnotify.ErrClosed
+	}
+	if err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return rw.fsnotifyWatcher.Add(path)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Remove recursively removes a directory tree from the list of watched files.
+func (rw *RecursiveWatcher) Remove(path string) error {
+	if rw.closed {
+		return nil
+	}
+	if err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return rw.fsnotifyWatcher.Remove(path)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Events returns the events channel.
+func (rw *RecursiveWatcher) Events() <-chan fsnotify.Event {
+	return rw.events
+}
+
+// Errors returns the errors channel.
+func (rw *RecursiveWatcher) Errors() <-chan error {
+	return rw.errors
+}
+
+// Close closes the RecursiveWatcher.
+func (rw *RecursiveWatcher) Close() error {
+	rw.closed = true
+	close(rw.done)
+	return rw.fsnotifyWatcher.Close()
+}
+
+func (rw *RecursiveWatcher) start() {
+	for {
+		select {
+		case event := <-rw.fsnotifyWatcher.Events:
+			info, err := os.Stat(event.Name)
```

Comment:
should we ignore events of type `Chmod`? just to reduce the amount of times we are calling `Stat`?

Reply:
or - if this is just a thin recursive wrapper of `fsnotify`, maybe we should only call `Stat` if the event type is `Create` or `Remove`? Doesn't seem like we need `Stat()` if the event type is something else

Reply:
Now that you say this I think we also should remove something from the watcher on `Rename`. Which means that we should just skip handling events for `Chmod`. The rest of the cases `Stat` is needed because we want to either update the directories we watch, or send the event message if it's not a directory.